### ISSUE TITLE
automatically archive old blueprints and inventory collections

### ISSUE THREAD

**Issue by davepacheco** [oxidecomputer/omicron#7278]
The basic plan here (from [this comment below](https://github.com/oxidecomputer/omicron/issues/7278#issuecomment-3237943644)):

- have Nexus create a Reconfigurator state file whenever the system runs the planner or manually sets a blueprint target.  This will include the planning input and inventory collection, etc. that were used for the planning.
- have sled agent archive these into the debug dataset
- update `reconfigurator-cli` to be able to load a whole bunch of these
- prune old blueprints from the database (keep only the most recent N)

That work is covered by these PRs:

- [ ] oxidecomputer/omicron-debug-dropbox#1
- [ ] #10306 
- [ ] #10308 
- [ ] #10307 
- [ ] #10309 

At various points, this issue has covered various other pieces of related work (e.g., updating our manual procedures).  What follow is **historical**.

---

tl;dr: I propose that we:

- Update our manual software update procedure to use `omdb db reconfigurator-save` before mupdate for every release.  Maybe we could store these into a debug dataset on the Scrimlet, sort of like a log file?  (Maybe we could just put them into a directory that already gets archived for log files?)
- When doing automated update, do the equivalent thing: serialize all the Reconfigurator state as JSON and store it into a debug dataset that gets archived.
- Optionally include these in support bundles, similar to whatever we're going to do with log files and core files.
- Consider: at the start of each update, both during manual mupdates and during automated update: delete all blueprints older than the current target or whose parent blueprint no longer exists.

---

Why: when a blueprint is replaced as the current target, it's no longer useful to the system.  But it never gets deleted unless an operator explicitly does so.  This wouldn't be a big deal (it's true for many other database records, too) except keeping old blueprints around in CockroachDB makes it much harder for us to evolve the system because the schema needs to be able to represent those old blueprints.  For example, if we want to add some new blueprint field in release N and it's always filled in by release N+1, if we were still keeping around blueprints from release N-1, we have to represent blueprints that don't have the field set, even though that's illegal in the software today.  The same also applies to inventory collections, but they get deleted automatically after a few minutes.

At the same time, historical blueprints and collections can be useful for understanding how a system has changed over time.  In debugging tricky path-dependent problems in production systems we might well want to go look at very old blueprints and collections.

Also:

- We always need to be able to read the current target blueprint.  When we come up after an update, the target blueprint was written by a previous version of the software.  So we're still somewhat constrained here, but we only have to support blueprints going back one release.
- It would be nice to always be able to read the last inventory collected by the previous release, though it might also be reasonable to say that we'll just always collect a new one after the update anyway.
- We also need to figure out what to do with blueprints that were never made the target.  We could remove them if their parent is removed because that means they can never be made the target again?

With the proposal above:

- we should have a historical record of all blueprints and many collections
- the only blueprint we necessarily keep across an upgrade is the target from before the upgrade
- we do need to keep supporting older blueprints for one release
- we might want to do that for inventory collections, too
- we never need to support reading blueprints or inventory collections from more than one release ago
- by using `reconfigurator-save`, we'll wind up with a bunch of other related useful state (e.g., inventory collections that went into some of those blueprints)

**Comment by jgallagher** [comment]
> * Consider: at the start of each update, both during manual mupdates and during automated update: delete all blueprints older than the current target or whose parent blueprint no longer exists.

I think this could happen _after_ updates too, right? If I have a system on release N that has 10 blueprints, 1 of which is the current target, and I'm going to update to release N+1, I have to still be able to read the existing current target, so presumably I can still read the other 9 too. Once I've made a new blueprint from release N+1, then I can delete all 10 of the old ones in one go.

I'm not sure there's a meaningful technical difference between these, but "don't delete old stuff until we're running new stuff" and "delete all the old stuff at once instead of deleting most of it before the upgrade and the last of it after" both seem appealing.

**Comment by davepacheco** [comment]
Yeah, maybe doing it both before and after is best.  Doing it before feels like it gives us a bit of a safety net if anything during or immediately after the upgrade goes wrong.  But I can see the appeal of having that information from after, too.

**Comment by davepacheco** [comment]
During today's meeting we agreed to go here directionally.  There's some question about details like:

- exactly when do we do the pruning (on the old version? on the new version?)
- do we want to make sure we create a target blueprint on each version (probably)
- let's make sure we store the `reconfigurator-save` output somewhere

**Comment by davepacheco** [comment]
The update to the manual procedure is being done here: https://github.com/oxidecomputer/customer-support/pull/287

We can leave this ticket open to cover doing the same during automatic update.

**Comment by davepacheco** [comment]
@jgallagher and I today discussed a more continuous, asynchronous process for archiving both blueprints and inventory collections.  One straw man: we have a new background task that's constantly slurping out both inventory collections and blueprints into some spot in the filesystem that gets archived by the global zone archiver into the debug dataset.  The background task also deletes inventory collections and blueprints when they've been archived _and_ it knows they're no longer needed (or useful).  We could have some heuristics to keep them around when they might be useful, like "if a blueprint is less than 48 hours old" or something.

Reading this issue again, I realize that this would not cover the planning inputs, which would be useful for debugging.  So another approach here would be to have the planner background task save the planning input, the new blueprint, and the inventory collection that it used into the debug dataset.  (Or maybe something like a whole reconfigurator state file, but limiting inventory and blueprints only to the collection it used and the blueprint that it generated.)  We could have the API that `omdb nexus blueprints regenerate` uses do the same thing.  This way, we'd _synchronously_ create an archive of the planning input, blueprint, and inventory collection any time we generate a new blueprint.  We'd still need a background task to clear these out of the database, but it wouldn't do the archiving itself.

With either of these designs, though, we still have the #3860 problem: if we expunge the Nexus zone that created the blueprint before the GZ archiver has saved it, then we'll lose the data.
